### PR TITLE
feat: wire zero-copy constants through level-1 pass to hip-compiler

### PR DESCRIPTION
**Pull request by BoarQing**
## Motivation

Handle external constants in the MLIR compilation pipeline by resolving them directly from the `location`/`offset`/`size` attributes on `onnx.Constant` ops, rather than wiring constant references through multiple layers of the compilation stack.

### Core: resolve external constants via `location` attribute (OnnxToHip.cpp)

- Add `resolveExternalLocationConstant()` to handle `onnx.Constant` ops carrying `location`/`offset`/`size` attributes (external data form emitted by the MLIR graph builder)
- The `offset` attribute holds the raw memory address; `size` holds the byte count
- Constants are either externalized to the `constants.bin` sidecar or inlined as `arith.constant`
- Update `lowerOnnxConstants` decision logic: try `value` attr first, then `location` attr, then error

## Net diff

12 files changed, 79 insertions, 119 deletions.

**Dependency:** https://github.com/ROCm/MorphiZen/pull/198

### CODE REVIEW

File: lib/Conversion/OnnxToHip/OnnxToHip.cpp [hunk 1]
```diff
@@ -154,22 +157,224 @@ struct ExternalizationState {
...
+  while (remaining > 0) {
+    size_t toWrite = std::min(remaining, bufSize);
+    extState->writer->fwrite(buf.data(), toWrite);
+    remaining -= toWrite;
+  }
+  extState->currentOffset += byteSize;
+  extState->constantSizes.push_back(byteSize);
+  extState->constantOffsets.push_back(entryOffset);
+
+  auto memrefType =
```
Comment (fhanuman) [inline lib/Conversion/OnnxToHip/OnnxToHip.cpp]: We are duplicating a lot of the code between `externalizeConstant` method and in this function, can we pull out this into a utility?

File: lib/Conversion/OnnxToHip/OnnxToHip.cpp [hunk 2]
```diff
@@ -154,22 +157,224 @@ struct ExternalizationState {
...
+      builder.getUnitAttr(), nullptr);
+  constOp->getResult(0).replaceAllUsesWith(toTensor.getResult());
+  constOp->erase();
+  ++extState->constantIndex;
+}
+
+/// Resolve a zero-copy constant: look up the initializer name in the
+/// constant data map, then externalize or inline depending on extState.
+static mlir::LogicalResult
+resolveZeroCopyConstant(mlir::ModuleOp module, mlir::Operation *constOp,
```
Comment (fhanuman) [inline lib/Conversion/OnnxToHip/OnnxToHip.cpp]: From the code it's not clear how the IR will look with the new changes, can you please add some IR snippets with the new changes for readability?

**Review by fhanuman — CHANGES_REQUESTED** [review]
Changes overall look fine to me; however, I want to look at how the IR looks after this change. Can you please dump IR and paste it in the description, so I can take a look? Please take a look at the following PR description for reference:

https://github.com/ROCm/onnx-hipdnn-ep/pull/74

**Comment by BoarQing** [comment]
## Perf Comparison: Single Layer Model (`single_layer_seq1.onnx`, 30s, MorphiZen EP)

| Metric | `main` | Zero-Copy PR | Delta |
|---|---|---|---|
| Session creation | 32.59 s | 22.11 s | **-32%** |
| First inference | 2985 ms | 268 ms | **-91%** |
| Avg latency | 10.36 ms | 14.54 ms | +40% |
| Throughput | 96.29 inf/s | 68.61 inf/s | -29% |
| P50 latency | 10 ms | 10 ms | — |
| P90 latency | 10 ms | 20 ms | +100% |
| P95 latency | 10 ms | 20 ms | +100% |
| P99 latency | 10 ms | 20 ms | +100% |
| Min latency | 10 ms | 10 ms | — |
| Max latency | 30 ms | 30 ms | — |
| Total runs | 2897 | 2064 | -29% |
| **Peak memory** | **5.77 GB** | **2.42 GB** | **-58%** |
| Avg CPU usage | 5% | 4% | — |

**Comment by BoarQing** [comment]
@fhanuman I have updated the PR. I have removed a few unnecessary changes introduced by design update.
I will address the throughout degradation during the weekend.

**Comment by BoarQing** [comment]
## Full Model Perf (`sequnece length = 1`, Zero-Copy PR, MorphiZen EP)
Note: this is on Strix.
| Metric | Value |
|---|---|
| Session creation | 361.7 s (~6 min) |
| First inference | 51,558 ms (~51.6 s) |
| Total inference runs | 1 |
| Avg latency | 42,461 ms (~42.5 s) |
| Throughput | 0.024 inf/s |
| Peak memory | ~23.6 GB |
| CPU usage | 35% |

**Comment by fhanuman** [comment]
> ## Full Model Perf (`sequnece length = 1`, Zero-Copy PR, MorphiZen EP)
> Metric	Value
> Session creation	361.7 s (~6 min)
> First inference	51,558 ms (~51.6 s)
> Total inference runs	1
> Avg latency	42,461 ms (~42.5 s)
> Throughput	0.024 inf/s
> Peak memory	~23.6 GB
> CPU usage	35%

Do we know why model compilation now takes `361` seconds? This is a significant regression compared to what we had. We were taking around `65` seconds.

https://amd.atlassian.net/wiki/spaces/AISW/pages/1574472924/Full+Llama+model+benchmarking

**Comment by BoarQing** [comment]
> > ## Full Model Perf (`sequnece length = 1`, Zero-Copy PR, MorphiZen EP)
> > Metric	Value
> > Session creation	361.7 s (~6 min)
> > First inference	51,558 ms (~51.6 s)
> > Total inference runs	1
> > Avg latency	42,461 ms (~42.5 s)
> > Throughput	0.024 inf/s
> > Peak memory	~23.6 GB
> > CPU usage	35%
> 
> Do we know why model compilation now takes `361` seconds? This is a significant regression compared to what we had. We were taking around `65` seconds.
> 
> https://amd.atlassian.net/wiki/spaces/AISW/pages/1574472924/Full+Llama+model+benchmarking

This is my local machine. Is has less cores. I will borrow Strix Halo after the debugging and get a valid comparsion

**Comment by fhanuman** [comment]
> ## Perf Comparison: Single Layer Model (`single_layer_seq1.onnx`, 30s, MorphiZen EP)
> Metric	`main`	Zero-Copy PR	Delta
> Session creation	32.59 s	22.11 s	**-32%**
> First inference	2985 ms	268 ms	**-91%**
> Avg latency	10.36 ms	14.54 ms	+40%
> Throughput	96.29 inf/s	68.61 inf/s	-29%
> P50 latency	10 ms	10 ms	—
> P90 latency	10 ms	20 ms	+100%
> P95 latency	10 ms	20 ms	+100%
> P99 latency	10 ms	20 ms	+100%
> Min latency	10 ms	10 ms	—
> Max latency	30 ms	30 ms	—
> Total runs	2897	2064	-29%
> **Peak memory**	**5.77 GB**	**2.42 GB**	**-58%**
> Avg CPU usage	5%	4%	—

Also, it's not clear why this change should affect performance. Does it make sense to provide a flag or environment variable to disable the zero-copy path? If this is going to impact performance, it would be good to have a mechanism to disable it and benchmark. I'm not suggesting we disable the feature by default — the flag itself can be off by default, keeping zero-copy enabled.

**Comment by BoarQing** [comment]
> > ## Perf Comparison: Single Layer Model (`single_layer_seq1.onnx`, 30s, MorphiZen EP)
> > Metric	`main`	Zero-Copy PR	Delta
> > Session creation	32.59 s	22.11 s	**-32%**
> > First inference	2985 ms	268 ms	**-91%**
> > Avg latency	10.36 ms	14.54 ms	+40%
> > Throughput	96.29 inf/s	68.61 inf/s	-29%
> > P50 latency	10 ms	10 ms	—
> > P90 latency	10 ms	20 ms	+100%
> > P95 latency	10 ms	20 ms	+100%
> > P99 latency	10 ms	20 ms	+100%
> > Min latency	10 ms	10 ms	—
> > Max latency	30 ms	30 ms	—
> > Total runs	2897	2064	-29%
> > **Peak memory**	**5.77 GB**	**2.42 GB**	**-58%**
> > Avg CPU usage	5%	4%	—
> 
> Also, it's not clear why this change should affect performance. Does it make sense to provide a flag or environment variable to disable the zero-copy path? If this is going to impact performance, it would be good to have a mechanism to disable it and benchmark. I'm not suggesting we disable the feature by default — the flag itself can be off by default, keeping zero-copy enabled.

@fhanuman We can set "XLNX_ABI_2_0_CLONE_EXTERNAL_DATA_THRESHOLD=999999999" to disable the offloading. This is a feature already done at MorphiZen repo. I think the performance data I have shown is related to my machine. I will try to get a correct comparison result tomorrow using the public Strix halo machine.

**Comment by BoarQing** [comment]
```
Session creation time cost: 101.06 s
First inference time cost: 1244 ms
Total inference time cost: 30.57 s
Total inference requests: 31
Average inference time cost total: 986.259865 ms
Total inference run time: 30.57 s
Number of inferences per second: 1.01
Avg CPU usage: 3 %
Peak working set size: 16452272128 bytes
Avg CPU usage:3
Peak working set size:16452272128
Runs:31
Min Latency: 0.97 s
Max Latency: 1.01 s
P50 Latency: 0.99 s
P90 Latency: 1.00 s
P95 Latency: 1.00 s
P99 Latency: 1.01 s
P999 Latency: 1.01 
```

**Comment by amd-mingw** [comment]
Hi @BoarQing , this is the result in my sharelab machine(10.228.227.228):
1. latest main(commit: 9fff3bd3e4c67fff1c10a4cae440a9eda82ac351)
image

2. latest main + morphizen commit(061a17c513905fa8c7ce707dfd625518bb68bbf0), only update morphizen/PR201
image

3. this PR, PR80
image

from the result, the performance improved, not degradation. and the init time reduced, too, memory reduced, too.
please double check it, or do a clean build as https://github.com/ROCm/onnx-hipdnn-ep/blob/main/docs/quick_start.md.
cc @fhanuman

File: lib/Conversion/OnnxToHip/OnnxToHip.cpp [hunk 3]
```diff
@@ -154,22 +155,209 @@ struct ExternalizationState {
...
-/// Large constants (at or above threshold, non-splat) are externalized:
-/// their raw data is appended to the sidecar binary via \p extState, and
-/// they are replaced by an extern memref.global (no initial value) carrying
-/// a hip.external_data {offset, size} attribute, plus a
-/// memref.get_global + bufferization.to_tensor bridge at the use site.
+/// Output IR when externalization is disabled (extState == nullptr):
+///
+///   %cst = arith.constant dense<[[1.0, 2.0, 3.0, 4.0], ...]>
+///       : tensor<2x4xf32>
+static mlir::LogicalResult
```
Comment (fhanuman) [inline lib/Conversion/OnnxToHip/OnnxToHip.cpp:320]: Could we please add some `lit` tests to lock down the `zero-copy` constants changes?

**Review by fhanuman — CHANGES_REQUESTED** [review]

File: lib/Conversion/OnnxToHip/OnnxToHip.cpp (hunk 3 above)
Comment (BoarQing) [reply lib/Conversion/OnnxToHip/OnnxToHip.cpp:320]: sure

**Comment by BoarQing** [comment]
> Hi @BoarQing , this is the result in my sharelab machine(10.228.227.228):
> 
> 1. latest main(commit: [9fff3bd](https://github.com/ROCm/onnx-hipdnn-ep/commit/9fff3bd3e4c67fff1c10a4cae440a9eda82ac351))
> 
> image
> 2. latest main + morphizen commit(061a17c513905fa8c7ce707dfd625518bb68bbf0), only update morphizen/PR201
> 
> image
> 3. this PR, PR80
> 
> image
> from the result, the performance improved, not degradation. and the init time reduced, too, memory reduced, too. please double check it, or do a clean build as https://github.com/ROCm/onnx-hipdnn-ep/blob/main/docs/quick_start.md. cc @fhanuman

Yes. I believe my slow compilation time caused by the architecture difference.

File: lib/Conversion/OnnxToHip/OnnxToHip.cpp (hunk 3 above)
Comment (BoarQing) [reply lib/Conversion/OnnxToHip/OnnxToHip.cpp:320]: I have added lit tests as u requested

**Comment by fhanuman** [comment]
> > Hi @BoarQing , this is the result in my sharelab machine(10.228.227.228):
> > 
> > 1. latest main(commit: [9fff3bd](https://github.com/ROCm/onnx-hipdnn-ep/commit/9fff3bd3e4c67fff1c10a4cae440a9eda82ac351))
> > 
> > image
> > 2. latest main + morphizen commit(061a17c513905fa8c7ce707dfd625518bb68bbf0), only update morphizen/PR201
> > image
> > 3. this PR, PR80
> > image
> > from the result, the performance improved, not degradation. and the init time reduced, too, memory reduced, too. please double check it, or do a clean build as https://github.com/ROCm/onnx-hipdnn-ep/blob/main/docs/quick_start.md. cc @fhanuman
> 
> Yes. I believe my slow compilation time caused by the architecture difference.

Hi @BoarQing, Just want to make sure I understand — is this change increasing model compilation / ORT session creation time? If so, that seems worth investigating, since the whole point of the zero-copy approach was to reduce compile time?

**Comment by BoarQing** [comment]
> > > Hi @BoarQing , this is the result in my sharelab machine(10.228.227.228):
> > > 
> > > 1. latest main(commit: [9fff3bd](https://github.com/ROCm/onnx-hipdnn-ep/commit/9fff3bd3e4c67fff1c10a4cae440a9eda82ac351))
> > > 
> > > image
> > > 2. latest main + morphizen commit(061a17c513905fa8c7ce707dfd625518bb68bbf0), only update morphizen/PR201
> > > image
> > > 3. this PR, PR80
> > > image
> > > from the result, the performance improved, not degradation. and the init time reduced, too, memory reduced, too. please double check it, or do a clean build as https://github.com/ROCm/onnx-hipdnn-ep/blob/main/docs/quick_start.md. cc @fhanuman
> > 
> > 
> > Yes. I believe my slow compilation time caused by the architecture difference.
> 
> Hi @BoarQing, Just want to make sure I understand — is this change increasing model compilation / ORT session creation time? If so, that seems worth investigating, since the whole point of the zero-copy approach was to reduce compile time?

No. Comparing to your last performance boost, this PR further reduced the compile time by 27s and increase inference per second by 1.4.

The figure 1 is main. The figure 2 added ur optimization at Morphizen. The figure 3 is my PR based on ur optimization.

File: test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// Test: onnx.Constant with location attribute but missing offset/size.
+//
+// The zero-copy path expects both offset and size integer attributes
+// alongside the location sentinel. Verify the diagnostic when they
+// are absent.
+//===----------------------------------------------------------------------===//
+
+// RUN: not hip-mlir-opt --hip-add-context-arg --convert-onnx-to-hip %s 2>&1 | FileCheck %s
+
+// CHECK: error: onnx.Constant with location attribute missing offset or size
```
Comment (fhanuman) [inline test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir:14]: These are all negative test cases, can we also please add a positive test point?

**Review by fhanuman — APPROVED** [review]

File: test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir (hunk 4 above)
Comment (BoarQing) [reply test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir:14]: > These are all negative test cases, can we also please add a positive test point?

Hi, I think test/lit/Conversion/onnx-to-hip/test_constant_externalize_named.mlir is the positive one

Comment (fhanuman) [reply test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir:14]: > > These are all negative test cases, can we also please add a positive test point?
> 
> Hi, I think test/lit/Conversion/onnx-to-hip/test_constant_externalize_named.mlir is the positive one

Yeah, that positive test point doesn't exercise your `zero-copy` path right?  If it takes a lot of effort, you can go ahead with the merge and this can be follow-up task too.

Comment (BoarQing) [reply test/lit/Conversion/onnx-to-hip/test_constant_location_missing_attrs.mlir:14]: > > > These are all negative test cases, can we also please add a positive test point?
> > 
> > 
> > Hi, I think test/lit/Conversion/onnx-to-hip/test_constant_externalize_named.mlir is the positive one
> 
> Yeah, that positive test point doesn't exercise your `zero-copy` path right? If it takes a lot of effort, you can go ahead with the merge and this can be follow-up task too.

sure. I will keep that in mind and provide a new test case for that path
